### PR TITLE
Generate fluentclient for subresource

### PR DESCRIPTION
**Pull request by junchuanwang**
Done
(1) Nested sub-resource interface for same namespace && their implementation
(2) Separate sub-resource interface for different namespace && their implementation
(3) AssociationKey binding method (getter) for sub-resources
(4) ComplexResourceKey binding method (getter) sub-resources
(5) Added tests

Couldn't/Woudln't do:
(1) Cannot get the nested indentation to work in Velocity template

### CODE REVIEW

**Review by karthikbalasub — COMMENTED** [review]
We also need to provide a link to the subresource from the parent resources's client. For example:

ParentClient {
   SubResourceClient getSubResource(Long parentId);
}

And then the SubResourceClient would just have to accept its own ids. Without the parentKey, the requests will not work. 

Please add integration tests for this.

**Comment by junchuanwang** [comment]
> We also need to provide a link to the subresource from the parent resources's client. For example:
> 
> ParentClient {
> SubResourceClient getSubResource(Long parentId);
> }
> 
> And then the SubResourceClient would just have to accept its own ids. Without the parentKey, the requests will not work.
> 
> Please add integration tests for this.

@karthikbalasub  Right, that is something I will add. This is basically the PathKeyBinding.

File: restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.greeting.restspec.json
Comment (karthikbalasub) [inline restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.greeting.restspec.json:89]: doc should be updated

File: restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.greeting.restspec.json
Comment (karthikbalasub) [inline restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.greeting.restspec.json:137]: same here. 

I think two levels is good enough for tests. Do we really need this deep scenario tested? 
I don't think even the second level is going to be used in the real world.

Comment (junchuanwang) [reply restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.greeting.restspec.json:137]: I am ok with not adding them, but I have tested these by checking out the result from this for simple keys, compound key and complex key.

Comment (karthikbalasub) [reply restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.greeting.restspec.json:137]: Since you've already added the resources, it is okay to keep these

File: restli-int-test-server/src/main/java/com/linkedin/restli/examples/greetings/server/AssociationsAssociationsResource.java
Comment (junchuanwang) [reply restli-int-test-server/src/main/java/com/linkedin/restli/examples/greetings/server/AssociationsAssociationsResource.java:19]: removed

File: restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java
Comment (junchuanwang) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java:662]: added

File: restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java
Comment (karthikbalasub) [inline restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java:766]: "Sub resource call without path key should fail."

File: restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java
Comment (karthikbalasub) [inline restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java:741]: Can you also add a test for getting the sub resource client using:

parentClient.subResourceOf(parentId)

Comment (junchuanwang) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java:741]: @karthikbalasub I didn't plan to add it here.

I had a separate task in the backlog for "wiring the interface", and I will add test there.

File: restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java
Comment (junchuanwang) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/TestParseqBasedFluentClientApi.java]: resolved

File: restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/ComplexKeySpec.java
Comment (karthikbalasub) [inline restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/ComplexKeySpec.java:55]: If you are using the short name, then you should also update the map to avoid future conflicts?

Comment (junchuanwang) [reply restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/ComplexKeySpec.java:54]: update the `checkIfShortNameConflictWithImports` so it has side effect for updating

File: restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.associations.restspec.json
Comment (BrianPin) [inline restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.associations.restspec.json:87]: nit: the first sub resource is with camel case, I think it might be better read if we make it same pattern?
`associationsassociations` maybe to `associationsAnotherSub` or whatever makes consistent.

File: restli-int-test-server/src/main/java/com/linkedin/restli/examples/greetings/server/AssociationsAssociationsResource.java
Comment (BrianPin) [inline restli-int-test-server/src/main/java/com/linkedin/restli/examples/greetings/server/AssociationsAssociationsResource.java:63]: Hmm, I think this is an example, so it is nit, I would suggest to rename from `AssociationsAssociationsResource` to something else. Maybe `AnotherAssociationResource`? Anyway, it may not be better, but the idea would be to have better example naming, I think the Greetings related name would make a good example name. It is your call.
And quite frankly, the parent resource naming is not very ideal too.

File: restli-int-test-server/src/main/java/com/linkedin/restli/examples/greetings/server/CollectionOfCollectionOfCollectionOfSimpleResource.java
Comment (BrianPin) [inline restli-int-test-server/src/main/java/com/linkedin/restli/examples/greetings/server/CollectionOfCollectionOfCollectionOfSimpleResource.java:56]: Hmm I think I see your naming pattern here. 
I think the pros are that it can be told there a number of fold of collections in the collections.
the cons are it didn't really easy to be found after a few months.

**Review by BrianPin — COMMENTED** [review]
Just reviewed half. Will finish the other half the other day.

**Review by karthikbalasub — APPROVED** [review]
Resolve Brian's comments before merging.

File: restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.associations.restspec.json
Comment (junchuanwang) [reply restli-int-test-api/src/main/idl/com.linkedin.restli.examples.greetings.client.associations.restspec.json:87]: change to associationsAssociations

File: restli-int-test-server/src/main/java/com/linkedin/restli/examples/greetings/server/AssociationsAssociationsResource.java
Comment (junchuanwang) [reply restli-int-test-server/src/main/java/com/linkedin/restli/examples/greetings/server/AssociationsAssociationsResource.java:63]: Won't change since the naming suppose to tell the parental relationship.

File: restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/FluentApiGenerator.java
Comment (BrianPin) [inline restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/FluentApiGenerator.java:208]: I know var `parentList` means the current parent list without include the current resource spec, but keep adding and removing makes me wonder if it is worth to consider this way (without keep adding and removing)
```
if there is sub resource in this spec:
  create a new list of parent spec by parentList.append(spec)
  and then call the recursive function.
```

**Review by BrianPin — APPROVED** [review]
Review finished, generally look good to me
consider my comment if possible.

**Pull request merged**
